### DISCUSSION TITLE
push-notifications.nextcloud.com is required for mobile phone?

### DISCUSSION THREAD

**Discussion by harihegen** [nextcloud/all-in-one#1707] [Questions] [answered]
Hello,

i have the AIO fully installed and external signalling, high performance backend, TURN and STUN servers all point to my server. Nevertheless, when chatting on the mobile phone, a push request is sent to Nextcloud. If I block nextcloud.com then the push to the mobile device no longer works either. What can that be, I thought if you host your own high-performance backend then you send the push yourself independently of the manufacturer nextcloud.

**Comment by small1** [comment]
This is how push on android and ios works. It is mostly sent through a proxy that then sends it via google or apple services and pushes it to the phone.  

The high performance backend for talk does signaling on chats only. It has nothing to do with push notifications. 

There are ways of turning this off and use other push services like unified push: https://apps.nextcloud.com/apps/uppush

**Reply by harihegen** [reply]
Thanks for the explanation, I'll take a look at the Unified Push

**Comment by nickvergessen** [comment, marked as answer]
As the little one already mentioned that is not possible at the moment.
The mobile apps people install are signed by a google/apple developer key. The same key needs to be used to sign the notifications. We can not ship that key with the Nextcloud server as otherwise everyone would have our developer key and could claim to be us.
Those services are not made for something like a federated universe but still assume there is a single entity behind them like with all the other services.

What we therefore did is creating the push proxy push-notifications.nextcloud.com, but to protect our users and their data we did the following:

* The Nextcloud server only knows a "proxy ID", it encrypts the content of the push notifications with a key of the mobile device and sends the notifications to the proxy.
* The proxy maps the the proxy ID to the push token of google/apple, but since it does not have the device key it can not access the content of the notification. It then signs it with our developer certificate and forwards it to apple/google accordingly.
* Google/Apply only see the push token + signed encrypted push notification, they do not know which Nextcloud server sent the notification neither the content. After verifying our signature they forward the encrypted message.
* The device has the key and can decrypt the notification to get the actual content.

**Reply by harihegen** [reply]
Thank you for the explanation, then I just have to live with it, I wanted to avoid the third party being able to record the metadata

**Reply by szaimen** [reply]
Here is also now a link to the documentation: https://github.com/nextcloud/notifications/blob/master/docs/push-v2.md#introduction

**Reply by harihegen** [reply]
> Here is also now ...

Thank you very much for this detailed explanation!
